### PR TITLE
[VM] Add e2e test for changing publishing option.

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

This PR will add the changing publishing option script be one of the whitelisted script at genesis. This will allow us to change the publishing option for VM in testnet.

Add an e2e test for changing publishing option. The test logic is as following:
1. At genesis the publishing option was set to default, thus only the allowed scripts can be executed and committed. We will submit a script with empty body to make sure it is rejected by the network.
2. Have the association submit a transaction to change the publishing option to `CustomScript`. Make sure a reconfiguration event happens and a new epoch is triggered.
3. Submits the exact same transaction again to the network and make sure that transaction can now be committed.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Related PRs

#2932

### CODE REVIEW

**Comment by sblackshear** [comment]
Very cool!

A question, though: would it make sense to add a test for updating the whitelist (i.e., adding a new script)?

File: language/stdlib/transaction_scripts/modify_publishing_option.move [hunk 1]
```diff
@@ -0,0 +1,5 @@
+use 0x0::ScriptWhitelist;
+
+fun main(args: vector<u8>) {
```
Comment (sblackshear) [inline language/stdlib/transaction_scripts/modify_publishing_option.move]: Comment explaining what this script does and who has the authority to use it?

Comment (runtian-zhou) [reply language/stdlib/transaction_scripts/modify_publishing_option.move]: > Very cool!
> 
> A question, though: would it make sense to add a test for updating the whitelist (i.e., adding a new script)?

In this current test, I changed the option from whitelist mode to the custom script mode. Adding a script should be fine, the main problem is just how to obtain the raw bytes of the new script. Do you want me to add a test for adding empty main script to the whitelist?

Comment (sblackshear) [reply language/stdlib/transaction_scripts/modify_publishing_option.move]: >  the main problem is just how to obtain the raw bytes of the new script. 

Ah, gotcha! I suppose you could have a link to the path of one of the `.mv`'s in `language/stdlib` and then hash that + use the script binary in your test to make sure both stay in sync.

I think it's important to have this test given that updating the whitelist is what we actually plan to do in prod.

Also, wondering about the raw bytes/Move value for on-chain configs question again... It seems like this might be easier to test if the whitelist was a Move value.

File: execution/executor/tests/storage_integration_test.rs [hunk 2]
```diff
@@ -241,6 +241,226 @@ fn test_reconfiguration() {
     // validator set fails to parse
 }
 
+#[test]
```
Comment (davidiw) [inline execution/executor/tests/storage_integration_test.rs]: We should probably have a dedicated directory for testing end to end between transactions, storage, and VM and probably outside the scope of the executor itself. Since these aren't really testing the executor.

Comment (sblackshear) [reply execution/executor/tests/storage_integration_test.rs]: Agreed. `libra/testsuite/tests/libratest` might be a good candidate. It already has a few tests that fit this description

Some of the tests in `libra/language/e2e-tests` could probably move to this new dir as well.

File: language/stdlib/transaction_scripts/modify_publishing_option.move (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/transaction_scripts/modify_publishing_option.move]: > Ah, gotcha! I suppose you could have a link to the path of one of the .mv's in language/stdlib and then hash that + use the script binary in your test to make sure both stay in sync.

Let me backtrack on this approach--https://github.com/libra/libra/pull/3090 will insist that all of the `.mv`'s on disk are in the whitelist.

File: execution/executor/tests/storage_integration_test.rs [hunk 3]
```diff
@@ -247,6 +247,520 @@ fn test_reconfiguration() {
...
+
+    let txn3 = get_test_signed_transaction(
+        validator_account,
+        /* sequence_number = */ 0,
+        validator_privkey.clone(),
+        validator_pubkey.clone(),
+        Some(script2.clone()),
+    );
+
+    // Create a dummy block prologue transaction that will bump the timer.
```
Comment (sblackshear) [inline execution/executor/tests/storage_integration_test.rs]: Why is this needed?

Comment (runtian-zhou) [reply execution/executor/tests/storage_integration_test.rs]: The executor is spawn up from the genesis (with a reconfiguration event in it). We need the timer to be bumped in order to trigger a new reconfiguration event!

File: execution/executor/tests/storage_integration_test.rs [hunk 4]
```diff
@@ -247,6 +247,520 @@ fn test_reconfiguration() {
...
+
+    // Transaction 3 is NOT committed.
+    let (t2, _) = response_items
+        .pop()
+        .unwrap()
+        .into_get_account_txn_by_seq_num_response()
+        .unwrap();
+    assert!(t2.is_none());
+}
+
```
Comment (sblackshear) [inline execution/executor/tests/storage_integration_test.rs]: These tests are pretty long/hard to follow... would it be possible to abstract some of the common setup into a helper method and/or to add helper methods for the common pattern of execute/commit/check responses?

Comment (sblackshear) [reply execution/executor/tests/storage_integration_test.rs]: It would be fine to do this in a follow-up too, but it seems like a very important problem to solve sometime soon. We don't want to add 500+ lines of tests for every new on-chain config...

Comment (runtian-zhou) [reply execution/executor/tests/storage_integration_test.rs]: cc @msmouse. Do you have any suggestions?
